core/src/kura.rs

```diff
@@ -4,258 +4,475 @@
-    #[iroha_futures::telemetry_future]
-    #[log("INFO", skip(self, block))]
-    pub async fn store(
+    fn write_block_index(&mut self, block_height: u64, start: u64, length: u64) -> Result<()> {
+        let mut index_file = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        let start_location = block_height * 2 * 8;
+        if start_location + 2 * 8 > index_file.metadata()?.len() {
+            index_file.set_len(start_location + 2 * 8)?;
+        }
+        index_file.seek(SeekFrom::Start(start_location))?;
+        // block0       | block1
+        // start, length| start, length  ... et cetera.
+        index_file.write_u64::<LittleEndian>(start)?;
+        index_file.write_u64::<LittleEndian>(length)?;
+        Ok(())
+    }
+
+    #[allow(clippy::integer_division)]
+    fn read_index_count(&self) -> Result<u64> {
+        let index_file = std::fs::OpenOptions::new()
+            .read(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        Ok(index_file.metadata()?.len() / (2 * 8)) // Each entry is 16 bytes.
+    }
+
+    fn write_index_count(&mut self, new_count: u64) -> Result<()> {
+        let index_file = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        let new_byte_size = new_count * 2 * 8;
+        index_file.set_len(new_byte_size)?;
+        Ok(())
+    }
+
+    fn read_block_data(
+        &self,
+        start_location_in_data_file: u64,
+        dest_buffer: &mut [u8],
+    ) -> Result<()> {
+        let mut data_file = std::fs::OpenOptions::new()
+            .read(true)
+            .open(self.path_to_blockchain.join("blocks.data"))?;
+        data_file.seek(SeekFrom::Start(start_location_in_data_file))?;
+        data_file.read_exact(dest_buffer)?;
+        Ok(())
+    }
+
+    fn write_block_data(
         &mut self,
-        block: VersionedCommittedBlock,
-    ) -> Result<HashOf<VersionedCommittedBlock>> {
-        match self.block_store.write(&block).await {
-            Ok(block_hash) => {
-                self.merkle_tree.add(block_hash);
-                self.broker.issue_send(ContinueSync).await;
-                Ok(block_hash)
-            }
-            Err(error) => {
-                let blocks = self
-                    .block_store
-                    .read_all()
-                    .await?
-                    .try_collect::<Vec<_>>()
-                    .await?;
-                self.merkle_tree = blocks
-                    .iter()
-                    .map(VersionedCommittedBlock::hash)
-                    .collect::<MerkleTree<_>>();
-                Err(error)
-            }
+        start_location_in_data_file: u64,
+        block_data: &[u8],
+    ) -> Result<()> {
+        let mut data_file = std::fs::OpenOptions::new()
```

Comment:
is this necessary considering that you have `create_files_if_they_do_not_exist`?